docs/wip/exchange-onboarding.md

```diff
@@ -0,0 +1,352 @@
+
+   * [Communication](#communication)
+   * [Requirements](#requirements)
+      * [Nix](#nix)
+         * [Optional: Enable IOHK's binary cache](#optional-enable-iohks-binary-cache)
+      * [Miscellaneous Utilities](#miscellaneous-utilities)
+   * [Wallet](#wallet)
+      * [Backup local state](#backup-local-state)
+      * [Fetch latest code](#fetch-latest-code)
+      * [Generate custom configuration](#generate-custom-configuration)
+      * [Build and run in the nix store](#build-and-run-in-the-nix-store)
+      * [Build and run a docker image](#build-and-run-a-docker-image)
+   * [Migrating from V0 to V1 API](#migrating-from-v0-to-v1-api)
+   * [Usage FAQs](#usage-faqs)
+      * [What are recommended hardware/software requirements for exchange wallets?](#what-are-recommended-hardwaresoftware-requirements-for-exchange-wallets)
+      * [How do I customize the wallet configuration?](#how-do-i-customize-the-wallet-configuration)
+      * [How do I export the CA certificate for the API?](#how-do-i-export-the-ca-certificate-for-the-api)
+      * [How do I know when the wallet has fetched all the blocks?](#how-do-i-know-when-the-wallet-has-fetched-all-the-blocks)
+      * [Creating a New Wallet](#creating-a-new-wallet)
+      * [Receiving Money](#receiving-money)
+      * [Sending Money](#sending-money)
+      * [Where can I find the API documentation?](#where-can-i-find-the-api-documentation)
+      * [How can I inspect runtime metrics and statistics?](#how-can-i-inspect-runtime-metrics-and-statistics)
+
+# Communication
+
+* An e-mail list needs to be provided by the exchange for IOHK to broadcast issues and upddates.
+* A slack room will be created for each exchange for direct communication with IOHK team.
+
+# Requirements
+
+## Nix
+
+The wallet is built using [nix package manager](https://nixos.org/nix/). To install it on
+most Linux distros download and run the installation script.
+
+    curl https://nixos.org/nix/install > install-nix.sh
+    . install-nix.sh
+
+Follow the directions and then log out and back in.
+
+### Optional: Enable IOHK's binary cache
+
+Skip this section if you prefer to build all code from IOHK
+locally. When the binary cache is enabled build steps will tend
+go faster.
+
+    sudo mkdir -p /etc/nix
+    cat <<EOF | sudo tee /etc/nix/nix.conf
+    binary-caches            = https://cache.nixos.org https://hydra.iohk.io
+    binary-cache-public-keys = hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ=
+    EOF
+
+## Miscellaneous Utilities
+
+Use `nix` to install essential utilities.
+
+    nix-env -i git
+
+# Wallet
+
+## Backup local state
+
+Skip to the next section if this is target machine doesn't yet have
+`cardano-sl` set up.
+
+To avoid catastrophic data loss, stop the wallet and backup the
+local databases, keys, certificates, and logs. By default, the
+local state will be in `./state-wallet-mainnet`, but may be
+elsewhere (see `stateDir` attribute in `./custom-wallet-config.nix`).
```

Comment:
Had breakage. See the nix file changes for docker that resolved those.

Reply:
So all the custom attrs work now in this branch with Docker?

Reply:
I just tried changing the ports, and it breaks in docker. Should we try to parse the ports used in ExposePorts, or just specify the ports have to remain the same in the docker container?

Reply:
Does it work if you change the port mapping specified in the `docker run` options? If so, let's just mention it near the `docker run` call; otherwise, I think it would be better to extract the appropriate attr from the custom config and parse it.

Reply:
Ping

Reply:
Sam just told me the port mapping didn't work, even when `-p` was manually and explicitly passed to `docker run`.

Reply:
Per Jake's comments just now, I'm going to convert these listen's to an attr set with host and port so we can specify exposePort of just the port portion. This also is how nixos handles this in most of it's modules, so this is a good best practice to do regardless.

Reply:
Port mappings work, just binding to localhost does not for obvious reasons.

I've resolved this with a NOTE added to sample-wallet-config.nix specifying that bind address should be 0.0.0.0 for docker users and they have to specify `-p` when running to expose on a custom port. We can revisit later if we want with getting the port from the sample config and exposing that, and maybe doing an assertion that IP is 0.0.0.0 when building docker images.

Reply:
> I'm going to convert these listen's to an attr set with host and port so we can specify exposePort of just the port portion

Was this done? If so, please update the `sample-wallet-config.nix` to reflect that.

Reply:
Which attrs from the `custom-wallet-config.nix` have you verified are supported in Docker? Please test any remaining and convey the lack of support in the doc.

Reply:
All of the attributes are supported. Tested `additionalNodeArgs` with `--json-log` and verified it generated the new file. Tested `ghcRuntimeArgs` with `-s` and verified extra output when node shuts down.

Since I was able to get it working with `-p` and using `0.0.0.0` as the bind IP, I didn't redo the config to split out host and port in the config. This could be done if we wanted to, but I'm not sure it's required anymore. There is a comment in sample wallet config that specifies with docker, you should use `0.0.0.0` as the node IP.